skimage/morphology/extrema.py

```diff
@@ -207,152 +210,360 @@ def h_minima(image, h, selem=None):
+           [0, 1, 1, 0, 2, 2, 0],
+           [0, 1, 1, 0, 2, 2, 0],
+           [1, 0, 0, 0, 0, 0, 3]])
+
+    Find local maxima by comparing to all neighboring pixels (maximal
+    connectivity):
+
+    >>> local_maxima(image)
+    array([[0, 0, 0, 0, 0, 0, 0],
+           [0, 1, 1, 0, 0, 0, 0],
+           [0, 1, 1, 0, 0, 0, 0],
+           [1, 0, 0, 0, 0, 0, 1]], dtype=uint8)
+    >>> local_maxima(image, indices=True)
+    (array([1, 1, 2, 2, 3, 3]), array([1, 2, 1, 2, 0, 6]))
+
+    Find local maxima without comparing to diagonal pixels (connectivity 1):
+
+    >>> local_maxima(image, selem=1)
+    array([[0, 0, 0, 0, 0, 0, 0],
+           [0, 1, 1, 0, 1, 1, 0],
+           [0, 1, 1, 0, 1, 1, 0],
+           [1, 0, 0, 0, 0, 0, 1]], dtype=uint8)
+
+    and exclude maxima that border the image edge:
+
+    >>> local_maxima(image, selem=1, include_border=False)
+    array([[0, 0, 0, 0, 0, 0, 0],
+           [0, 1, 1, 0, 1, 1, 0],
+           [0, 1, 1, 0, 1, 1, 0],
+           [0, 0, 0, 0, 0, 0, 0]], dtype=uint8)
     """
-    # find the minimal grey level difference
-    h = _find_min_diff(image)
-    if h == 0:
-        return np.zeros(image.shape, np.uint8)
-    if not np.issubdtype(image.dtype, np.floating):
-        h = 1
-    local_max = h_maxima(image, h, selem=selem)
-    return local_max
+    image = np.asarray(image)
+    if image.size == 0:
+        # Return early for empty input
+        return np.array([], dtype=(np.intp if indices else np.uint8))
+
+    if include_border:
+        # Ensure that local maxima are always at least one smaller sample away
+        # from the image border
+        image = _fast_pad(image, image.min())
+
+    if selem is None:
+        # Choose maximal connectivity if not provided
+        selem = image.ndim
+    if np.isscalar(selem):
+        selem = ndi.generate_binary_structure(image.ndim, selem)
+    else:
+        # Validate custom structured element
+        selem = np.asarray(selem, dtype=bool)
+        # Must specify neighbors for all dimensions
+        if selem.ndim != image.ndim:
+            raise ValueError("selem and image must have the same number of "
+                             "dimensions")
+        # Must only specify direct neighbors
+        if any(s != 3 for s in selem.shape):
+            raise ValueError("dimension size in selem is not 3")
+
+    neighbor_offsets = _offset_to_raveled_neighbours(image.shape, selem)
+
+    # Array of flags used to store the state of each pixel during evaluation.
+    # Possible states are:
+    #   3 - first or last value in a dimension
+    #   2 - potentially part of a maximum
+    #   1 - evaluated and part of a maximum
+    flags = np.zeros(image.shape, dtype=np.uint8)
+    _set_edge_values_inplace(flags, value=3)
+
+    # Up-cast dtype of image for wrapped Cython function
+    if np.issubdtype(image.dtype, np.unsignedinteger):
+        dtype = np.uint64
+    elif np.issubdtype(image.dtype, np.integer):
+        dtype = np.int64
```

Comment:
Is there a specific reason why you didn't include all numeric types in the fused types? ie why should I need to multiply my volume by 8x to find maxima in a uint8 image? (Which is probably *the* most common use case for this function?)

Reply:
I could do that. But I thought this would increase the source of the generated binary significantly. If that is acceptable I'd be happy to change this because I agree that upcasting is not a very elegant solution to cover all dtypes.

Reply:
We are talking about 1.4 MiB (all dtypes) vs 1.2 MiB (3 dtypes) here. Furthermore I just noticed that the dtype numpy.float16 has no equivalent numpy.float16_t in Cythons PXD file for NumPy. I wonder why that is. This might be a problem if we want to support that type.

Reply:
I'm not worried about the 200KB. A 1000x1000 uint8 image upcasting will already end up using an extra 7MB of RAM if we don't include all the types.

float16 is rare so I reckon it can be ignored/upcasted if needed.